crates/jose/src/claims.rs

```diff
@@ -235,6 +239,82 @@ impl From<&TimeOptions> for TimeNotBefore {
     }
 }
 
+/// Hash the given token with the given algorithm for an ID Token claim.
+///
+/// Returns `None` if the algorithm is [`JsonWebSignatureAlg::None`].
+///
+/// According to the [OpenID Connect Core 1.0 specification].
+///
+/// [OpenID Connect Core 1.0 specification]: https://openid.net/specs/openid-connect-core-1_0.html#CodeIDToken
+pub fn hash_token(alg: JsonWebSignatureAlg, token: &str) -> anyhow::Result<Option<String>> {
+    let bits = match alg {
+        JsonWebSignatureAlg::Hs256
+        | JsonWebSignatureAlg::Rs256
+        | JsonWebSignatureAlg::Es256
+        | JsonWebSignatureAlg::Ps256
+        | JsonWebSignatureAlg::Es256K => {
+            let mut hasher = Sha256::new();
+            hasher.update(token);
+            let hash = hasher.finalize();
+            // Left-most half
+            hash.get(..16).map(ToOwned::to_owned)
+        }
+        JsonWebSignatureAlg::Hs384
+        | JsonWebSignatureAlg::Rs384
+        | JsonWebSignatureAlg::Es384
+        | JsonWebSignatureAlg::Ps384 => {
+            let mut hasher = Sha384::new();
+            hasher.update(token);
+            let hash = hasher.finalize();
+            // Left-most half
+            hash.get(..24).map(ToOwned::to_owned)
+        }
+        JsonWebSignatureAlg::Hs512
+        | JsonWebSignatureAlg::Rs512
+        | JsonWebSignatureAlg::Es512
+        | JsonWebSignatureAlg::Ps512 => {
+            let mut hasher = Sha512::new();
+            hasher.update(token);
+            let hash = hasher.finalize();
+            // Left-most half
+            hash.get(..32).map(ToOwned::to_owned)
+        }
+        JsonWebSignatureAlg::EdDsa => return Err(anyhow::anyhow!("unsupported algorithm")),
+        JsonWebSignatureAlg::None => return Ok(None),
```

Comment:
I think I'd return an error here instead. You're throwing an error anyway later if it returns None.